fuzzy-message-detection/src/combiner.rs

```diff
@@ -0,0 +1,61 @@
+use alloc::vec::Vec;
+
+/// Combines multiple filtered messages into a smaller collection
+/// of messages.
+pub struct FilterCombiner;
+
+impl FilterCombiner {
+    /// Combines messages whose flags have been filtered ([detected](crate::FmdScheme::detect)) with
+    /// different detection keys.
+    pub fn combine<T: Eq + Clone>(filtered_messages: &[Vec<T>]) -> Option<Vec<T>> {
+        let intersect_all = filtered_messages.iter().cloned().reduce(|acc, filter| {
+            let (shortest, largest) = if acc.len() < filter.len() {
+                (acc, filter)
+            } else {
+                (filter, acc)
+            };
+            let mut intersection = Vec::<T>::with_capacity(shortest.len());
+            // Below runs in Ω(|shortest|²) time. An alternative
+            // is `Hashset.intersection` from std lib.
+            for element in shortest {
+                if largest.contains(&element) {
+                    intersection.push(element);
+                }
+            }
+            intersection
+        });
+
+        if intersect_all.is_some() && intersect_all.clone().unwrap().is_empty() {
+            return None;
+        }
+
+        intersect_all
+    }
```

Comment:
```suggestion
    pub fn combine<T: Eq + Clone>(filtered_messages: &[vec::Vec<T>]) -> vec::Vec<T> {
        if filtered_messages.is_empty() {
            return vec![];
        }

        let mut result = vec![];
        filtered_messages[0].iter().for_each(|msg| {
            if filtered_messages[1..]
                .iter()
                .all(|list| list.contains(&msg))
            {
                result.push(msg.clone());
            }
        });

        result
    }
```

Swapping the shortest and largest elements doesn't seem to offer benefits and instead introduces additional copies. This suggestion aims to make the code cleaner and reduce unnecessary clones. Please ignore if not applicable.

Reply:
It may be better to return an empty vector instead of None if there is no intersection?